hack/dind-cluster.sh

```diff
@@ -127,12 +126,84 @@ cluster's rc file to configure the bash environment:
   fi
 }
 
+function add-node () {
+  local config_root=$1
+  local deployed_config_root=$2
+  local cluster_id=$3
+  local wait_for_cluster=$4
+
+  echo "Adding node to dind cluster '${cluster_id}'"
+
+  # Error if a cluster is not already configured
+  check-containers "add-node"
+
+  # Find the first free number
+  local first_free=1
+  for cid in $( ${DOCKER_CMD} ps -a --filter "name=${NODE_PREFIX}" --format "{{.Names}}" | sort -V ); do
+    if [[ "${cid}" != "${NODE_PREFIX}${first_free}" ]]; then
+      break
+    fi
+    (( first_free += 1 ))
+  done
+  local node_name="${NODE_PREFIX}${first_free}"
+
+  start-container "${config_root}" "${deployed_config_root}" "${NODE_IMAGE}" "${node_name}"
+  echo
+  echo "Added node '${node_name}'"
+
+  if [[ -n "${wait_for_cluster}" ]]; then
+    wait-for-cluster "${config_root}" "$(count-nodes)"
+  fi
+}
+
+function start-container() {
+  local config_root=$1
+  local deployed_config_root=$2
+  local image=$3
+  local name=$4
+
+  local volumes run_cmd
+  volumes=""
+
+  ${DOCKER_CMD} run -dt ${volumes} -v "${config_root}:${deployed_config_root}" --privileged \
+                --name="${name}" --hostname="${name}" "${image}" > /dev/null
+}
+
+function delete-node () {
+  local config_root=$1
+  local cluster_id=$2
+  local node_name=$3
+
+  echo "Removing node '${node_name}' from dind cluster '${cluster_id}'"
+  sudo echo -n
```

Comment:
Because the state can get into an interesting and hard-to-recover from place if you stop the docker processes and then abort the cleanup that requires sudo.